payment/p2tr.go

```diff
@@ -0,0 +1,170 @@
+package payment
+
+import (
+	"errors"
+
+	"github.com/btcsuite/btcd/btcec/v2"
+	"github.com/btcsuite/btcd/btcec/v2/schnorr"
+	"github.com/btcsuite/btcd/chaincfg/chainhash"
+	"github.com/vulpemventures/go-elements/address"
+	"github.com/vulpemventures/go-elements/network"
+	"github.com/vulpemventures/go-elements/taproot"
+)
+
+const (
+	segwitVersion = byte(0x01)
+)
+
+// TaprootPaymentData is included in Payment struct to store Taproot-related data
+type TaprootPaymentData struct {
+	XOnlyTweakedKey    []byte
+	XOnlyInternalKey   []byte
+	RootScriptTreeHash *chainhash.Hash
+	ScriptTree         *taproot.IndexedElementsTapScriptTree
+}
+
+// FromTweakedKey creates a P2TR payment from a tweaked output key
+func FromTweakedKey(
+	tweakedKey *btcec.PublicKey,
+	net *network.Network,
+	blindingKey *btcec.PublicKey,
+) (*Payment, error) {
+	if tweakedKey == nil {
+		return nil, errors.New("tweaked key can't be empty or nil")
+	}
+
+	if net == nil {
+		net = &network.Liquid
+	}
+
+	return &Payment{
+		Network:     net,
+		BlindingKey: blindingKey,
+		Taproot: &TaprootPaymentData{
+			XOnlyTweakedKey: schnorr.SerializePubKey(tweakedKey),
+		},
+	}, nil
+}
+
+func FromTaprootScriptTreeHash(
+	internalKey *btcec.PublicKey,
+	rootHash *chainhash.Hash,
+	net *network.Network,
+	blindingKey *btcec.PublicKey,
+) (*Payment, error) {
+	if internalKey == nil {
+		return nil, errors.New("internal key can't be empty or nil")
+	}
+
+	if net == nil {
+		net = &network.Liquid
+	}
+
+	return &Payment{
+		Network:     net,
+		BlindingKey: blindingKey,
+		Taproot: &TaprootPaymentData{
+			XOnlyInternalKey:   schnorr.SerializePubKey(internalKey),
+			RootScriptTreeHash: rootHash,
+		},
+	}, nil
+}
+
+func FromTaprootScriptTree(
+	internalKey *btcec.PublicKey,
+	tree *taproot.IndexedElementsTapScriptTree,
+	net *network.Network,
+	blindingKey *btcec.PublicKey,
+) (*Payment, error) {
+	if internalKey == nil {
+		return nil, errors.New("internal key can't be empty or nil")
+	}
+
+	if net == nil {
+		net = &network.Liquid
+	}
+
+	return &Payment{
+		Network:     net,
+		BlindingKey: blindingKey,
+		Taproot: &TaprootPaymentData{
+			XOnlyInternalKey: schnorr.SerializePubKey(internalKey),
+			ScriptTree:       tree,
+		},
+	}, nil
+}
+
+func (p *Payment) taprootBech32() (*address.Bech32, error) {
+	payload := &address.Bech32{
+		Prefix:  p.Network.Bech32,
+		Version: segwitVersion,
+	}
+	if p.Taproot.XOnlyTweakedKey != nil {
```

Comment:
Before this, we must ensure that `p.Taproot != nil` otherwise we must return an error.
If the user attempts to generate a TR address from a non TR payments, we must prevent panics so to speak.